File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    maxi = float('-inf')
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        
        def dfs(root):
            if not root:
                return 0

            left = max(0, dfs(root.left))
            right = max(0, dfs(root.right))
            self.maxi = max(self.maxi, left + right + root.val)
            
            return max(left, right) + root.val
        
        dfs(root)
        return self.maxi

File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private int maxi;
    public int helperMaxPathSum(TreeNode root) {
        if(root == null) return 0;

        int left = Math.max(0, helperMaxPathSum(root.left));
        int right = Math.max(0, helperMaxPathSum(root.right));
        maxi = Math.max(maxi, left + right + root.val);

        return Math.max(left, right) + root.val;
    }

    public int maxPathSum(TreeNode root) {
        maxi = Integer.MIN_VALUE;
        helperMaxPathSum(root);
        return maxi;
    }
}